Add CNG backend and crypto/rsa to the user guide (#697)
microsoft/go: eng/doc/fips/UserGuide.md

      - [func Sum512](#func-sum512)
      - [func Sum512_224](#func-sum512_224)
      - [func Sum512_256](#func-sum512_256)
    - [crypto/rsa](#cryptorsa)
      - [func DecryptOAEP](#func-decryptoaep)
      - [func DecryptPKCS1v15](#func-decryptpkcs1v15)
      - [func DecryptPKCS1v15SessionKey](#func-decryptpkcs1v15sessionkey)
      - [func EncryptPKCS1v15](#func-encryptpkcs1v15)
      - [func SignPKCS1v15](#func-signpkcs1v15)
      - [func SignPSS](#func-signpss)
      - [func VerifyPKCS1v15](#func-verifypkcs1v15)
      - [func VerifyPSS](#func-verifypss)
      - [func GenerateKey](#func-generatekey-1)
      - [func GenerateMultiPrimeKey](#func-generatemultiprimekey)
      - [func PrivateKey.Decrypt](#func-privatekeydecrypt)
      - [func PrivateKey.Sign](#func-privatekeysign-1)
    - [crypto/subtle](#cryptosubtle)
    - [crypto/tls](#cryptotls)

## Using Go crypto APIs

This section describes how to use Go crypto APIs in a FIPS compliant manner.

As a general rule, crypto APIs will delegate low-level operations to the crypto backend if these rules are met:

- The operation is supported by the crypto backend.
- The set of input parameters are supported by the crypto backend.

> If any of the previous rules are not met, the operation will fall back to standard Go crypto unless otherwise specified.

Reviewer: I don't think this doc ever says outright what fallbacks mean for FIPS compliance... maybe add this:

```suggestion
If any of the previous rules are not met, the operation will fall back to standard Go crypto unless otherwise specified. Standard Go crypto will behave as expected but is not FIPS compliant.
```

      - [func Sum512](#func-sum512)
      - [func Sum512_224](#func-sum512_224)
      - [func Sum512_256](#func-sum512_256)
    - [crypto/rsa](#cryptorsa)
      - [func DecryptOAEP](#func-decryptoaep)
      - [func DecryptPKCS1v15](#func-decryptpkcs1v15)
      - [func DecryptPKCS1v15SessionKey](#func-decryptpkcs1v15sessionkey)
      - [func EncryptPKCS1v15](#func-encryptpkcs1v15)
      - [func SignPKCS1v15](#func-signpkcs1v15)
      - [func SignPSS](#func-signpss)
      - [func VerifyPKCS1v15](#func-verifypkcs1v15)
      - [func VerifyPSS](#func-verifypss)
      - [func GenerateKey](#func-generatekey-1)
      - [func GenerateMultiPrimeKey](#func-generatemultiprimekey)
      - [func PrivateKey.Decrypt](#func-privatekeydecrypt)
      - [func PrivateKey.Sign](#func-privatekeysign-1)
    - [crypto/subtle](#cryptosubtle)
    - [crypto/tls](#cryptotls)

## Using Go crypto APIs

This section describes how to use Go crypto APIs in a FIPS compliant manner.

> As a general rule, crypto APIs will delegate low-level operations to the crypto backend if these rules are meet:

Reviewer: ```suggestion
As a general rule, crypto APIs will delegate low-level operations to the crypto backend if these rules are met:
```

- The operation is supported by the crypto backend.
- The set of input parameters are supported by the crypto backend.

> If any of the previous rules are not meet, then the operation will fall back to standard Go crypto unless not otherwise specified.

Reviewer: ```suggestion
If any of the previous rules are not met, the operation will fall back to standard Go crypto unless otherwise specified.
```

NewCipher creates and returns a new [cipher.Block](https://pkg.go.dev/crypto/cipher#Block).

**Requirements**

- `key` length must be 16, 24, or 32 bytes.

**Implementation**

> <details><summary>OpenSSL</summary>

Reviewer: It might be a good idea to say in the intro section that these are clickable. Some people won't have seen these in a GitHub-rendered markdown document before and the arrows can be missed, or misinterpreted as fancy bullet points. Then it reads as "this function is implemented in OpenSSL and CNG".

> ![image](https://user-images.githubusercontent.com/12819531/184944273-78dc5041-f61a-4369-a3a5-bb04002827cb.png)

I usually include something like "(click me!)" in the summary text to make sure it's obvious, but maybe it's too much clutter:

> **Implementation**
> 
> <details><summary>OpenSSL (click for details)</summary></details>
> 
> <details><summary>CNG (click for details)</summary></details>

(When I started using sections in some .NET PR infrastructure I had to explain to someone once that it was clickable, and ever since I've always tried to include some kind of prompt. 😄)

Author: Good point. It is indeed confusing. I've added `(click for details)` on every expandable section, and it still looks good IMO. I prefer to be explicit here, as people would not expect that the prologue contains instructions on how to interact with the markdown.

> 

Reviewer: Looks like the code block's return part (`ctr cipher.BlockMode`) is wrong:

```
func cipher.NewCTR(block Block, iv []byte) Stream
```

(Can't comment on the line because it didn't actually change in this PR. 😄)

Author: Done!

</details>

#### func [NewOFB](https://pkg.go.dev/crypto/cipher#NewOFB)

cipher.NewOFB is not implemented by any backend.

#### func [StreamReader.Read](https://pkg.go.dev/crypto/cipher#StreamReader.Read)

```go
func (r cipher.StreamReader) Read(dst []byte) (n int, err error)
```

**Requirements**

- The CNG backend does not implement this function.
- `r.S` must be an object created by [cipher.NewCTR](#func-newctr).

#### func [StreamWriter.Write](https://pkg.go.dev/crypto/cipher#StreamWriter.Write)

```go
func (w cipher.StreamWriter) Write(src []byte) (n int, err error)
```

**Requirements**

- The CNG backend does not implement this function.
- `r.S` must be an object created by [cipher.NewCTR](#func-newctr).

#### func [StreamWriter.Close](https://pkg.go.dev/crypto/cipher#StreamWriter.Close)

```go
func (w cipher.StreamWriter) Close() error
```

Does not contain crypto algorithms, out of FIPS scope.

### [crypto/des](https://pkg.go.dev/crypto/des)

Not implemented by any backend.

### [crypto/dsa](https://pkg.go.dev/crypto/dsa)

Not implemented by any backend.

### [crypto/dsa](https://pkg.go.dev/crypto/dsa)

> Not implemented by any backend.

Reviewer: Noticed crypto/dsa is listed twice, also in the TOC.

Author: Done!

Sign signs `digest` with `priv`.

**Requirements**

> - `rand` must be boring.RandReader, else Sign will panic. `crypto/rand.Reader` normally meet this invariant as it is assigned to boring.RandReader in the crypto/rand init function.

Reviewer: ```suggestion
- `rand` must be boring.RandReader, else Sign will panic. `crypto/rand.Reader` normally meets this invariant as it is assigned to boring.RandReader in the crypto/rand init function.
```

Int returns a uniform random value in [0, max). It panics if max <= 0.

**Requirements**

> - `rand` must be boring.RandReader. `crypto/rand.Reader` normally meet this invariant as it is assigned to boring.RandReader in the crypto/rand init function.

Reviewer: ```suggestion
- `rand` must be boring.RandReader. `crypto/rand.Reader` normally meets this invariant as it is assigned to boring.RandReader in the crypto/rand init function.
```

> func Prime(rand io.Reader, bits int) (p *big.Int, err error)

Reviewer: Noticed this is a duplicate func signature rather than the func doc.

Author: Done!

**Requirements**

> `rand` must be boring.RandReader. `crypto/rand.Reader` normally meet this invariant as it is assigned to boring.RandReader in the crypto/rand init function.

Reviewer: ```suggestion
- `rand` must be boring.RandReader. `crypto/rand.Reader` normally meets this invariant as it is assigned to boring.RandReader in the crypto/rand init function.
```

#### func [Read](https://pkg.go.dev/crypto/rand#Read)

```go
func Read(b []byte) (n int, err error)
```

Read is a helper function that calls rand.Reader.Read using io.ReadFull.

**Requirements**

> - `rand.Reader` must be boring.RandReader.

Reviewer: Should this include something similar to the extra info from rand.Prime?

```suggestion
- `rand.Reader` must be boring.RandReader. This invariant is normally met as `rand.Reader` is assigned to boring.RandReader in the crypto/rand init function.
```

New returns a new hash.Hash computing the SHA1 checksum.

**Implementation**

<details><summary>OpenSSL</summary>

The hash is generated using [EVP_MD_CTX_new] and [EVP_DigestInit_ex] with the algorithm [EVP_sha1].

The hash.Hash methods are implemented as follows:

- `Write` using [EVP_DigestUpdate].

> - `Sum(` using [EVP_DigestFinal].

Reviewer: Looks like `Sum(` should be `Sum`, a few more times in the doc too.

Author: Done!

New224 returns a new hash.Hash computing the SHA224 checksum.

**Requirements**

- The CNG backend does not implement this function.

> <details><summary>OpenSSL (click for details)</summary>

Reviewer: ```suggestion
**Implementation**

<details><summary>OpenSSL (click for details)</summary>
```

**Requirements**

- `hash` can be one of the following values: crypto.MD5, crypto.MD5SHA1, crypto.SHA1, crypto.SHA224, crypto.SHA256, rypto.SHA384, or crypto.SHA512. Else VerifyPSS will fail.
- The CNG backend does not support crypto.MD5SHA1 nor crypto.SHA224.
- `digest` must be the result of hashing a message using a FIPS compliant hashing algorithm.
- `opts` can be nil.
- `opts.SaltLength` can either be a number of bytes, or one of the following constants: rsa.PSSSaltLengthAuto and rsa.PSSSaltLengthEqualsHash.
- The CNG backend does not support nil `opts` nor rsa.PSSSaltLengthAuto.

**Implementation**

<details><summary>OpenSSL (click for details)</summary>

`sig` is verified using using [EVP_PKEY_verify] with `RSA_PKCS1_PSS_PADDING` pad mode.

</details>

<details><summary>CNG (click for details)</summary>

`sig` is verified using [BCryptVerifySignature] with [PSS_PADDING_INFO] padding information and `BCRYPT_PAD_PSS` pad mode.

</details>

#### func [GenerateKey](https://pkg.go.dev/crypto/rsa#GenerateKey)

```go
func rsa.GenerateKey(rand io.Reader, bits int) (priv *rsa.PrivateKey, err error)
```

GenerateKey generates a public and private key pair.

**Requirements**

> - `rand` must be boring.RandReader. `crypto/rand.Reader` normally meet this invariant as it is assigned to boring.RandReader in the crypto/rand init function.

Reviewer: ```suggestion
- `rand` must be boring.RandReader. `crypto/rand.Reader` normally meets this invariant as it is assigned to boring.RandReader in the crypto/rand init function.
```

- `bits` must be either 2048 or 3072.

**Implementation**

<details><summary>OpenSSL (click for details)</summary>

`priv` is a wrapper around [EVP_PKEY] generated using [EVP_PKEY_keygen].

</details>

<details><summary>CNG (click for details)</summary>

`priv` is generated using [BCryptGenerateKeyPair] with the [algorithm identifier] `BCRYPT_RSA_ALGORITHM`.

</details>

#### func [GenerateMultiPrimeKey](https://pkg.go.dev/crypto/rsa#GenerateMultiPrimeKey)

```go
func rsa.GenerateMultiPrimeKey(rand io.Reader, nprimes int, bits int) (priv *rsa.PrivateKey, err error)
```

GenerateMultiPrimeKey generates a multi-prime RSA keypair of the given bit size.

**Requirements**

> - `rand` must be boring.RandReader. `crypto/rand.Reader` normally meet this invariant as it is assigned to boring.RandReader in the crypto/rand init function.

Reviewer: ```suggestion
- `rand` must be boring.RandReader. `crypto/rand.Reader` normally meets this invariant as it is assigned to boring.RandReader in the crypto/rand init function.
```

- If `opts` type is `*rsa.OAEPOptions`, it calls [rsa.DecryptOAEP](#func-decryptoaep)`(opts.Hash.New(), rand, priv, ciphertext, opts.Label)`.
- If `opts` type is `*rsa.PKCS1v15DecryptOptions` and `opts.SessionKeyLen > 0`, it calls [rsa.DecryptPKCS1v15SessionKey](#func-decryptpkcs1v15sessionkey)`(rand, priv, ciphertext, plaintext)` with a random `plaintext`.
- If `opts` type is `*rsa.PKCS1v15DecryptOptions` and `opts.SessionKeyLen == 0`, it calls [rsa.DecryptPKCS1v15](#func-decryptpkcs1v15)`(rand, priv, ciphertext)`.
- Else it returns an error.

Each case may impose additional parameter requirements. After determining which case applies, check the linked function to find the additional restrictions.

#### func [PrivateKey.Sign](https://pkg.go.dev/crypto/rsa#PrivateKey.Sign)

```go
func (priv *rsa.PrivateKey) Sign(rand io.Reader, digest []byte, opts crypto.SignerOpts) ([]byte, error)
```

Sign signs `digest` with `priv`.

The sign function depends on `opts`:

- If `opts` type is `*rsa.PSSOptions`, it calls [rsa.SignPSS](#func-signpss)`(rand, priv, pssOpts.Hash, digest, opts)`
- Else it calls [rsa.SignPKCS1v15](#func-signpkcs1v15)`(rand, priv, opts.HashFunc(), digest)`.

> Check the function named in the applicable case for the parameters' restrictions.

Reviewer: ```suggestion
Each case may impose additional parameter requirements. After determining which case applies, check the linked function to find the additional restrictions.
```

`priv` is generated using [BCryptGenerateKeyPair] with the [algorithm identifier] `BCRYPT_RSA_ALGORITHM`.

</details>

#### func [GenerateMultiPrimeKey](https://pkg.go.dev/crypto/rsa#GenerateMultiPrimeKey)

```go
func rsa.GenerateMultiPrimeKey(rand io.Reader, nprimes int, bits int) (priv *rsa.PrivateKey, err error)
```

GenerateMultiPrimeKey generates a multi-prime RSA keypair of the given bit size.

**Requirements**

- `rand` must be boring.RandReader. `crypto/rand.Reader` normally meet this invariant as it is assigned to boring.RandReader in the crypto/rand init function.
- `nprimes` must be 2. 
- `bits` must be either 2048 or 3072.

**Implementation**

<details><summary>OpenSSL</summary>

`priv` is a wrapper around [EVP_PKEY] generated using [EVP_PKEY_keygen].

</details>

<details><summary>CNG</summary>

`priv` is generated using [BCryptGenerateKeyPair] with the [algorithm identifier] `BCRYPT_RSA_ALGORITHM`.

</details>

#### func [PrivateKey.Decrypt](https://pkg.go.dev/crypto/rsa#PrivateKey.Decrypt)

```go
func (priv *PrivateKey) Decrypt(rand io.Reader, ciphertext []byte, opts crypto.DecrypterOpts) (plaintext []byte, err error)
```

Decrypt decrypts `ciphertext` with `priv`.

> The depcrypt function depends on `opts`:

Reviewer: ```suggestion
The decrypt function depends on `opts`:
```

- If `opts` is nil, it calls [rsa.DecryptPKCS1v15](#func-decryptpkcs1v15)`(rand, priv, ciphertext)`.
- If `opts` type is `*rsa.OAEPOptions`, it calls [rsa.DecryptOAEP](#func-decryptoaep)`(opts.Hash.New(), rand, priv, ciphertext, opts.Label)`.
- If `opts` type is `*rsa.PKCS1v15DecryptOptions` and `opts.SessionKeyLen > 0`, it calls [rsa.DecryptPKCS1v15SessionKey](#func-decryptpkcs1v15sessionkey)`(rand, priv, ciphertext, plaintext)` with a random `plaintext`.
- If `opts` type is `*rsa.PKCS1v15DecryptOptions` and `opts.SessionKeyLen == 0`, it calls [rsa.DecryptPKCS1v15](#func-decryptpkcs1v15)`(rand, priv, ciphertext)`.
- Else it returns an error.

> Check the function named in the applicable case for the parameters' restrictions.

Reviewer: I know I suggested this wording, 😄 but reading it again and maybe understanding it a bit better, I think it should be more specific that this is an action you need to take to make sure you're using the func right:

```suggestion
Each case may impose additional parameter requirements. After determining which case applies, check the linked function to find the additional restrictions.
```

  - `tls.ECDSAWithP521AndSHA512`

[EVP_EncryptUpdate]: https://www.openssl.org/docs/manmaster/man3/EVP_EncryptUpdate.html
[EVP_DecryptUpdate]: https://www.openssl.org/docs/manmaster/man3/EVP_DecryptUpdate.html

> [RAND_bytes]: https://www.openssl.org/docs/man3.0/man3/RAND_bytes.html

Reviewer: I noticed some of these are `manmaster` and some `man3.0`, is there a reason?

Author: I suppose Bing returns `manmaster` in some cases, but the good one is `man3.0`. Fixed!
